### ISSUE TITLE
TinyMCE integration

### ISSUE THREAD

**Issue by tx4x** [OneArmyWorld/onearmy#246]
hey guys,
here are the first caveat bumping up during my TinyMCE integration.

**Custom Image plugin (to use FirebaseFileUploader !?)**
- i had to copy over the sources of the [original image plugin](https://github.com/tinymce/tinymce/tree/master/src/plugins/image) because TinyMCE only takes an upload URL which won't work with the Firebase except you have a server which does the actual upload
- by copying the sources, it also requires to rewrite slightly the image plugin to meet the TS rules in place
- not only the image plugin has to be rewritten, it also comes with some dependencies which are then loaded twice. this can be prevented to build tinymce yourself which is needed anyways to get rid of the most built-in loaded code.

however, this penalties/overhead seems still ok to me and so I will go ahead for now.

Other issues : 
- editor controls such as the save button beneath the editor have to be done in higher component because for now I am only doing a React component which wraps TinyMCE and adds the needed plugins (image, file, whatever) and it takes a single property to setup all, ie: Editor variant={VARIANTS.SMALL|HOWTO|COMMENT|REPLY} handler={this:IEditorDelegate}

**Comment by tx4x** [comment]
update,
i've got the original image plugin ported over into the code base.
![image](https://user-images.githubusercontent.com/1205100/51983106-296aba00-2498-11e9-911c-0f717b4ca93c.png)

next:

I will adopt the image plugin to the code base's ui-widgets (Dialog, button, FirebaseImageUploader). this provides then also the required leg work for all follow up plugins (content/link/tag picker)

I will bump here Friday again
g

**Comment by BenGamma** [comment]
Thanks, looks promising 💪 
I reference this issue in #211

**Comment by chrismclarke** [comment]
@tx4x feel free to update the dependencies and make a separate PR if it helps you and you don't want to wait.

**Comment by chrismclarke** [comment]
Also really like the look of the image popup, will this also be triggered on image drag/drop or copy/paste? No biggie if not.

**Comment by tx4x** [comment]
@chrismclarke : actually this was the default image plugin, i am to replace this as said with : 

Dialog (popup) => material-core/dialog, currently done as reusable dialog, as core candidate
Upload button => FirebaseFileUploader
Ok Button => core/Button
Cancel Button => core/Button

about the logic: this won't behave exactly anymore as original tinymce image plugin but some remain intact: 
- double click on image : opens popup (to replace image)
- other behavior: let me know, delete is done within the editor (del/backspace) = no idea yet how to delete this in firebasedb)

anyway, i will know by tomorrow more, it's a little bumpy since i have to use react within the tinymce plugin infrastructure

**Comment by tx4x** [comment]
@chrismclarke : a few hours later,
here a version with the ui elements from /Step : 

![image](https://user-images.githubusercontent.com/1205100/52065501-ecc1c000-2576-11e9-9eea-73280381c080.png)

currently i am not sure about :
1. how to add the file drop zone
2. multiple uploads in one row, would love to have this
3. wrap FirebaseFileUploader in a way that it is showing the uploaded image in a preview
4. how to delete uploaded images (removal by user)
5. how to deal with oversize images
6. what image edit features (per variant), ie: width, height, alt text|title|caption I should enable

more or less i've got a confident handle on doing tinymce plugins

for now i will cleanup all, and notify you guys to rollout the next steps. i assume this will be about the content picker then. 

also, the editor is being pre-configured with the [variant attribute](https://github.com/tx4x/onearmy/blob/module--discussions/src/components/Editor/config.ts): POST, COMMENT, DISCUSSION which sets it all up, from minimal to advanced editing. 

see you in a bit :-)

**Comment by tx4x** [comment]
> will this also be triggered on image drag/drop or copy/paste? No biggie if not.

yes, that's possible, for now there is no routing, double click can open whatever you like, dragn drop is TinyMCE built-in and works flawless as expected, same for copy & paste

**Comment by tx4x** [comment]
variant = inline :

![image](https://user-images.githubusercontent.com/1205100/52066707-6064cc80-2579-11e9-9281-3216c2ae56c3.png)

editor controls/commands only showing when focused

**Comment by tx4x** [comment]
![image](https://user-images.githubusercontent.com/1205100/52067186-58595c80-257a-11e9-92a8-b0289463cc11.png)

**Comment by tx4x** [comment]
ah, i figured how to delete images from firedb, sorry

**Comment by tx4x** [comment]
@BenGamma or @chrismclarke : 
so I am ready for adding PP v4 specs (behaviour, presets). Mind the questions above...

![image](https://user-images.githubusercontent.com/1205100/52125841-02de8780-262e-11e9-8af4-
9327a67b4073.png)

It's working quite well and I managed to get the code in the way as I like, clean and less, see here for the code for above : https://github.com/tx4x/onearmy/blob/module--discussions/src/components/Editor/plugins/image/ui/Dialog.tsx

See you monday

**Comment by tx4x** [comment]
more @todos: 
- add resize recommendation/actions as gmail (large, medium, small) when pixel/mb limit exceeded
- clarify that original sizes should kept intact : should be
- enable multiple uploads : i want it, tired of clicking around if it's not needed :-)
- integrate preview for making next steps,  in discussion and how-to pages

**Comment by chrismclarke** [comment]
This is all looking really promising! I'm away travelling for the next week but if you reach any sticking points just send a message with a direct mention and I'll try to prioritise it. Otherwise can't wait to put it to use!

In terms of deleting image there's a few ways to do it and I'm not particularly set on current implementation vs api endpoint, so if you come across a way you think it would be useful to have things done then just let us know and I can add it to the roadmap.

**Comment by tx4x** [comment]
@chrismclarke : well, there is indeed something you may can enlighten me:

react-hooks (this what i use) requires currently whatever is in React@next right now. I have opened a pull-request for that but i am not sure we're on the same page about react-hooks. i guess the time v4 goes live, react-hooks should be ready for production, so we should be fine.

see why react hooks is the new way of doing FP like code : https://www.youtube.com/watch?v=wXLf18DsV-I 

g

**Comment by chrismclarke** [comment]
thanks @tx4x  for opening #252 for discussion on hooks, have added my comments there so we can keep this issue cleaner for TinyMCE

**Comment by tx4x** [comment]
@mattia-io @chrismclarke 

I need to know a few things to finish the editor component which is mainly preconfigured via presets [here](https://github.com/tx4x/onearmy/blob/module--discussions/src/components/Editor/config.ts)

The idea is to use the Editor via the property 'variant'. So I need to know what to disable for each module.

For instance : 
**module-disussion** : 

all base plugins,ie: undredo, tables, text formatting, insert image (s), mention, insert link, embedd media

**module-how-to**

step's description

base plugins : undredo, text formatting, insert link (other posts, howtows pickers)

question: no free inserting of pictures at any position, tables within the description ?

etc...etc..

Apart from the pre-configuration, these are still to do : 
1. basic localization for save/cancel
2. network link went down, save in localstorage til back ? 
3. where or how to show error messages (failed image uploads/removals)
4. focus management/policy for buttons or other widgets is missing
5. image oversize handling
6. image drop into editor
7. retrieve all users for the mentions plugin
8. retrieve all posts/how-to-pages for the link picker
9. multiple image upload
10. clarify max editor height / autoresize per variant
11. content parser for display only

**Comment by tx4x** [comment]
12. fully forgot: more likely it's better to make a custom build of Tinymce, someone please create a repository ? Currently the OneArmy app is already pretty heavy and the Tiny build defaults add around 1MB ...

**Comment by tx4x** [comment]
@mattia-io , @BenGamma 

a list of all modules would be already enough for me to keep going. Of course I checked but I couldn't find modules for :

1. polls (any user)
2. machine upgrades/addons
3. user profile/space
 
and so all i see now is :
- how-tow
- discussion
- comment
- post

may be someone can point me to the final design ? thanks
g

**Comment by BenGamma** [comment]
Hey @tx4x , I understand your excitement about the plateform and your contribution is helpfull but please try to keep issues a bit organize here.
You are throwing a few different things and it is hard to follow you.
Next time create a proper issue for the thing you want to discuss or it will be lost in an ocean of comments.

To come back to the TinyMCE integration :
* What is the undredo plugin you are talking about ?
* Can you clarify your question :

> no free inserting of pictures at any position, tables within the description ?

* We are keeping everything in this repository for now.
* Please understand that this project is still a massive work in project and that everything is not ready to build yet. If you cannot find a module it means that it is not ready for developement. For now if you want to keep on going, everything is in the issues list.

**Comment by tx4x** [comment]
ok, never mind, i will go ahead as it fits :-)
please check the available plugins for TinyMCE on their demo pages.
g

**Comment by chrismclarke** [comment]
Yeah that's quite a lot to get any head around! I've tried to group my responses by how I've interpreted the question.

1,6,9 - Features you're working on (assumed) -, great, keep us in the loop when things are done of if you need specific support
2,3,4 - Questions/decisions that haven't been made (assumed) - I would suggest implementing something simple in the first instance which we can later review and go back to the UI guys to see what improvements they suggest (as @BenGamma said above, some stuff in issues but still a long way to go on others)
5 - We have server-side image compression so assuming a non-issue.
7,8 - Retrieving data - Create a mock for now, add separate issue once done and I'll support best way to pull in what you need
10,11 - Not exactly sure what you mean, hopefully falls into one of above categories, if needs explicit input just let me know
12 - Created a fork which can be the base for a One Army custom editor here: https://github.com/OneArmyWorld/tinymce, although this will add considerable work as we would either need to add its build into the CI pipeline, or find a better way to host the dist version (e.g. publish to npm). I'll let you think about how you might want to do this and if required let's talk about in another thread.

I think before you try to take things too much further it might be good to put what you have into operation, and then base future development of a couple different threads (again as benji said, it's getting quite hard to fully follow this thread).

**Comment by tx4x** [comment]
ok, thanks for the directions, deal ! 

Give me some days to roll out the next slices
- PR to use editor in existing modules, controlled via presets set in the 'variant' prop
- PR for custom TinyMCE build, basically just a collection of long-term apply-able patches and scripts. i don't expect to have this changed often, only for major TinyMCE updates but that happens once in 2 years, no worry.
- single PRs and issues for all other points

I work in the next months half time on this and the other issues, the other half I am 5 meters away in the machine shop or just off time :-)
g

**Comment by tx4x** [comment]
@chrismclarke : i think you can delete the tinymce repo because adding a script which downloads via git the tinymce sources, then applies the patches (remove plugs,...) and builds it on the fly seems more easy. this will add then of course some dev deps but that's still ok. 
g

**Comment by chrismclarke** [comment]
If not publishing a package with the version we want would it not be easier to just build the version we want offline and stick the output dist directly in assets or somewhere?

**Comment by tx4x** [comment]
@chrismclarke , this is how it's done for now (official TinyMCE 4.9.3 build placed in public/vendor/tinymce) . I tried making a custom build yesterday but I came across various TS build errors, I suppose this is their way of pushing folks into their commercial version, at least this is the impression I've got looking at their github  activities.

However, I also made a loader for TinyMCE, which loads the script on demand (Editor:componentWillMount)

I'd say, lets forget about the tinymce build story altogether for now, there isn't much difference in script size to gain after all, and there is so much more work still : link/people/... picker, mention plugin etc.. 
g

**Comment by chrismclarke** [comment]
I think that sounds sensible for the time being. Agreed on priorities
